releng_sop/koji_sign.py

```diff
@@ -0,0 +1,715 @@
+        :rtype:  tuple
+        """
+        sigkeys = [i.lower() for i in sigkeys]
+
+        def _find_signed_rpm(rpm_info, matches_by_rpm_id):
+            for sigkey in sigkeys:
+                path = self._get_rpm_path(rpm_info, sigkey)
+                if os.path.isfile(path):
+                    matches_by_rpm_id[rpm_info["id"]] = True
+                    return
+            matches_by_rpm_id[rpm_info["id"]] = False
+
+        signed, unsigned = self._find_rpms(rpm_info_list, _find_signed_rpm)
+        assert len(rpm_info_list) == len(signed) + len(unsigned)
+        return signed, unsigned
+
+    def find_signed_rpms_in_main_copies(self, rpm_info_list, sigkeys):
+        """
+        Find signed RPMs in main copies.
+
+        Main (unsigned) copies of RPMs are typically unsigned.
+        They can be signed in rare situations such as imported signed 3rd party packages.
+        In such case re-use existing signature rather than sign the RPMs again.
+
+        :param rpm_info_list: List of koji rpm_info dictionaries
+        :type  rpm_info_list: list
+        :param sigkeys: List of sigkeys
+        :type  sigkeys: list
+        :return: (matched, unmatched) koji rpm_info dictionaries
+        :rtype:  tuple
+        """
+        sigkeys = [i.lower() for i in sigkeys]
+        signed = []
+        unsigned = []
+
+        def _find_signed_rpm_in_main_copies(rpm_info, matches_by_rpm_id):
+            path = self._get_rpm_path(rpm_info, None)
+            try:
+                _, sigkey = self._get_rpm_sighdr_sigkey(path)
+            except:
+                sigkey = None
+            if sigkey in sigkeys:
+                self.logger.debug("Found a main copy signed with '%s': %s" % (sigkey, path))
+                matches_by_rpm_id[rpm_info["id"]] = True
+            else:
+                self.logger.debug("Unsigned main copy: %s" % path)
+                matches_by_rpm_id[rpm_info["id"]] = False
+
+        signed, unsigned = self._find_rpms(rpm_info_list, _find_signed_rpm_in_main_copies)
+        return signed, unsigned
+
+    def write_signed_rpms_from_sigcache(self, rpm_info_list, sigkey, commit=False):
+        """
+        Reconstruct RPMs in koji from existing signed headers in sigcache.
+
+        :param rpm_info_list: List of koji rpm_info dictionaries
+        :type  rpm_info_list: list
+        :param sigkey: Sigkey
+        :type  sigkey: str
+        :param commit: Disable dry-run, apply changes for real.
+        :type  commit: bool=False
+        """
+        if commit:
+            self.koji_session.multicall = True
+
+        for rpm_info in rpm_info_list:
+            path = self._get_rpm_path(rpm_info, sigkey)
+
+            msg = "Writing RPM from '%s' sigcache: %s" % (sigkey, path)
+            self.log("info", msg, commit=commit)
+
+            if commit:
+                self.koji_session.writeSignedRPM(rpm_info, sigkey)
+
+        if commit:
+            self.koji_session.multiCall()
```

Comment:
It would be nice to do error checking here.
The ``writeSignedRPM`` could hit several issues (all raised as ``koji.GenericError``):
* ``koji.GenericError, "Not an internal rpm: %s (from %s)"``
* ``koji.GenericError, "No such path: %s"``
* ``koji.GenericError, "Not a regular file: %s"``
* ``koji.GenericError, "No cached signature for package %s, key %s"``